subscriber: add nested spans in json formatter

mockersf opened the pull request:
Fixes: #704

## Motivation

Nested spans are not available in json formatter

## Solution

I added a field `spans` in the json, with an array of the nested spans. I reused the span representation from the existing `span`, and pushed all the span representation in a `vec` using `ctx.visit_spans`. I didn't remove the `span` field as it can be useful to have direct access to the current span

hawkw commented:
The CI failure is unrelated (it's due to a flaky test in `tracing-appender`, which will be fixed by #737). I've gone ahead and re-run the failed jobs.

hawkw commented on tracing-subscriber/src/fmt/format/json.rs:
```diff
@@ -89,37 +120,21 @@ where
                 }
             }
 
+            let mut spans = vec![];
+            ctx.visit_spans(|span| {
+                spans.push(visit_span(span));
+                Ok(())
+            })?;
```
take it or leave it: it might be a bit more idiomatic to use the [`Context::scope`](https://docs.rs/tracing-subscriber/0.2.5/tracing_subscriber/layer/struct.Context.html#method.scope) iterator here instead. `visit_spans` is essentially an older version of `scope` and might be deprecated eventually.

```suggestion
            let spans = ctx.scope()
                .map(visit_span)
                .collect::<Vec<_>>();
```

mockersf replied:
Thanks, I switched to using `Context::scope`

hawkw commented on tracing-subscriber/src/fmt/format/json.rs:
```diff
@@ -89,37 +120,21 @@ where
                 }
             }
 
+            let mut spans = vec![];
+            ctx.visit_spans(|span| {
+                spans.push(visit_span(span));
+                Ok(())
+            })?;
+            if !spans.is_empty() {
+                serializer.serialize_entry("spans", &spans)?;
+            }
```
Rather than `collect`ing to a `Vec`, which allocates, it should be possible to serialize the spans without an intermediate allocation, using `serde`'s [`SerializeSeq`](https://docs.serde.rs/serde/ser/trait.SerializeSeq.html). This might offer a small performance improvement.

I think we could do this by adding a newtype which wraps the `Context::spans` iterator and implements `serde::Serialize` like this:

```rust
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: Serializer,
    {
         let mut seq = s.serialize_seq();
         for span in self.0 {
             // ...
         }
    }
```
and passing that struct to `serialize_entry` as the value.

Of course, this would involve a little more complexity than the current approach, so I'm fine with merging what we have now and doing this optimization as a follow-up. We might want to do some benchmarking to determine how much of a performance improvement we would get from it.

mockersf replied:
I implemented `Serialize` for newtypes for the context and the span, to keep it all managed by serde.
Please let me know if you see any issue in how I did it

hawkw commented:
Oh, a couple of API notes:

- Do you think we might want to add the option to turn off the `span` field that contains the current span? We probably want to continue generating it by default to avoid breaking code that parses the current output format, but because we are now also generating the `spans` list, some users might not want to duplicate the current span.
- Do we also want to provide the option to disable the `spans` list for users who _only_ want the immediate current span? Disabling the spans list would be a bit more efficient in cases where it isn't wanted.

What do you think?

mockersf commented:
Thank you for your review !
adding options to turn off current span or nested spans logging sounds good, I'll do it soon (hopefully sometime in the next 2 days)

mockersf commented:
Added two options to the json formatter:
- `disable_current_span` to not log the current span
- `disable_span_list` to not log the span list

hawkw commented on tracing-subscriber/src/fmt/format/json.rs:
```diff
@@ -76,6 +188,8 @@ where
         let meta = event.metadata();
 
         let flatten_event = self.format.flatten_event;
+        let disable_current_span = self.format.disable_current_span;
+        let disable_span_list = self.format.disable_span_list;
```
Why are these `let` bindings needed?

mockersf replied:
they are not, it was how it was already done: https://github.com/tokio-rs/tracing/blob/master/tracing-subscriber/src/fmt/format/json.rs#L78.
I removed them

hawkw commented on tracing-subscriber/src/fmt/format/mod.rs:
```diff
@@ -302,6 +302,26 @@ impl<T> Format<Json, T> {
+    /// See [`format::Json`](../fmt/format/struct.Json.html)
+    #[cfg(feature = "json")]
+    #[cfg_attr(docsrs, doc(cfg(feature = "json")))]
+    pub fn disable_current_span(mut self, disable_current_span: bool) -> Format<Json, T> {
+        self.format.disable_current_span(disable_current_span);
+        self
+    }
+
+    /// Use the full JSON format without the span list.
+    ///
+    /// See [`format::Json`](../fmt/format/struct.Json.html)
+    #[cfg(feature = "json")]
+    #[cfg_attr(docsrs, doc(cfg(feature = "json")))]
+    pub fn disable_span_list(mut self, disable_span_list: bool) -> Format<Json, T> {
```
Same naming suggestions as above.

hawkw commented:
@mockersf Is there anything I can do to help get this branch ready to merge? I think it's very close and just needs the handful of API tweaks I suggested in https://github.com/tokio-rs/tracing/pull/741#pullrequestreview-429370246.

mockersf commented:
Sorry for taking so long, I changed the naming according to your review.


I also fixed the fact that when there is no span, `"spans":[]` was added to the log but not `span` (now they are both not present)

hawkw commented on tracing-subscriber/src/fmt/format/json.rs:
```diff
@@ -34,19 +34,132 @@ use tracing_log::NormalizeEvent;
+        self.display_current_span = display_current_span;
+    }
+
+    /// If set to `false` event won't contain spans.
```
```suggestion
    /// If set to `false`, formatted events won't contain a list of all currently entered spans.
```

hawkw commented on tracing-subscriber/src/fmt/mod.rs:
```diff
@@ -555,6 +555,32 @@ impl<T, F, W> SubscriberBuilder<format::JsonFields, format::Format<format::Json,
+        }
+    }
+
+    /// Sets whether or not the JSON layer being built will include a list.
```
```suggestion
    /// Sets whether or not the JSON layer being built will include a list of all
    /// currently entered spans in formatted events.
```

hawkw commented on tracing-subscriber/src/fmt/mod.rs:
```diff
@@ -555,6 +555,32 @@ impl<T, F, W> SubscriberBuilder<format::JsonFields, format::Format<format::Json,
         }
     }
+
+    /// Sets whether or not the JSON layer being built will include the current span.
```
```suggestion
    /// Sets whether or not the JSON layer being built will include the current span
    /// in formatted events.
```

hawkw commented on tracing-subscriber/src/fmt/format/mod.rs:
```diff
@@ -302,6 +302,26 @@ impl<T> Format<Json, T> {
         self
     }
+
+    /// Use the full JSON format and sets whether or not it will include the current span.
```
```suggestion
    /// Sets whether or not the formatter will include the current span in
    /// formatted events.
```

hawkw commented on tracing-subscriber/src/fmt/format/mod.rs:
```diff
@@ -302,6 +302,26 @@ impl<T> Format<Json, T> {
+        self
+    }
+
+    /// Use the full JSON format sets whether or not it will include the span list.
```
```suggestion
    ///  Sets whether or not the formatter will include a list of all currently
    /// entered spans in formatted events.
```

mockersf commented:
Lint and doc should be good now 👍 

> > I also fixed the fact that when there is no span, `"spans":[]` was added to the log but not `span` (now they are both not present)
> 
> I'm not totally convinced about this --- I think it might be better to always include those fields but emit an empty list or `null` when there is no current span, so that code consuming the JSON can expect those fields to always be present when they are enabled. This isn't a strongly held opinion, though, so I'm also fine with merging it as is --- @LucioFranco, what do you think about this?

I did this to match the current behaviour, but it can be done either way. I have the change locally if you want, I won't take days this time 😄

hawkw commented:
@mockersf it looks like [rustfmt needs to be run again](https://github.com/tokio-rs/tracing/pull/741/checks?check_run_id=805583182). Once we take care of that and decide what the most correct way to handle no current span is, I think this is good to merge.

LucioFranco commented:
So chatting with @hawkw and @davidbarsky on discord led to a few ideas:

one is to go with the span as a list as you suggested @mockersf 

```json
"span": []
```

To me this seems reasonable but I think lists in json are a bit funky and all over the place for such a structured item like spans. What I was suggesting was using objects instead:

```json
"span": {
  "name": "MySpan",
  "parent": {}
}
```

This might be a bit more intuitive but is more verbose. So I wanted to hear from actual users what they prefer 😄

hawkw commented:
It would be nice to make a decision as to how we want to output the span hierarchy (as nested objects or as a list) so that we can move forward with merging this. Personally, I don't really care --- I would just like to ship support for nested spans in the JSON formatter in _some_ form.

I'm tagging @sazzer — since you opened #704, and are actually using the JSON formatter in production, I'm guessing you might have an opinion about this. Also @danburkert. If anyone else is using the JSON formatter and wants to weigh in, it would be great to get your thoughts as well!

sazzer commented:
@hawkw I'm not in Production yet, but I am using it.

Personally speaking I'm easy on how it works as long as it's consistent. I'd expect that a list would make more sense than an object. If it was an object I'm not sure exactly how it would be structured? The comment above from @LucioFranco seems to imply that each span would potentialy have a `parent` element, but that then means that if you get quite deep nesting you've got multiple levels to traverse. Essentially it would be:
```
{
  "name": "span_1",
  "parent": {
    "name": "span_2",
    "parent": {
      "name": "span_3",
      "parent": {
        "name": "span_4"
      }
    }
  }
}
```

whereas the same thing as a list would be:
```
[
  { "name": "span_1" },
  { "name": "span_2" },
  { "name": "span_3" },
  { "name": "span_4" }
]
```

Cheers

hawkw commented:
@sazzer yup, the examples you posted are more or less what it would look like. My two cents is that the nested object graph theoretically conveys the semantics more correctly, but that the list is probably easier to traverse and _definitely_ easier to read. I'm willing to merge this either way.

hawkw commented:
I guess another potential advantage of the nested form is that it offers less configuration surface area: rather than switching between a `"span"` field with a single object and a `"spans"` field that's a list, we always emit a `"span"` field, and the only switch to flip is whether or not that field contains the span's parent(s)...

sazzer commented:
One thing to consider is how they work with JSON friendly log tools.
We use ELK at work, where Logstash consumes JSON log files into Elastic search (from Java there - sorry!) And that's fantastic. There are plenty of other log management tools that natively consume JSON too, and I'm not sure how they would handle lists vs objects in this case.

I mention this because the name of the span seems something you'd want to filter on, and the other properties are important to easily pick out when diagnosing issues.

Cheers

neoeinstein commented:
Based on my understanding of how to go about querying in elastisearch and CloudWatch Insights, having a list is going to be a lot easier to deal with when looking for a particular span. I can just filter on `spans.name` rather than needing to look in `span.name` and `span.parent.name` and `span.parent.parent.name` or know its precise location in the hierarchy (which might be dependent on the configured logging level). This becomes more of a problem with arbitrary depth recursive structures.

Because the spans form a path, the parent-child relationship can be unambiguously determined from an array, and the searchability is less hindered. For that reason, I'd prefer to have spans be an array. We should just determine whether the root comes first or last.

hawkw commented:
Okay, it's sounding like the list is _probably_ better for most of the tools that will be ingesting these logs, and we should move forward with this PR as-is --- unless anyone else has a strong reason why the list is _not_ correct.

> Because the spans form a path, the parent-child relationship can be unambiguously determined from an array, and the searchability is less hindered. For that reason, I'd prefer to have spans be an array. We should just determine whether the root comes first or last.

Currently, it looks like the array is ordered so that the root span is the first element, and the current leaf span is the last. I'm fine with that, although I think ordering it in the other direction (leaf -> root) feels slightly more intuitive to me personally --- I don't _really_ care which we choose, as long as it's consistent. However, it looks like this isn't currently documented. @mockersf, can we add the ordering to the documentation for the span list, please?

mockersf commented:
I added the ordering and updated the example with how it will look with default options

hawkw commented:
It looks like this needs to be updated to compile with the changes from 50e857414c8b454b5b56b744f34f23271a2dfc87. Once that's taken care of, I'm happy to merge it, unless anyone else has any last issues.

hawkw commented:
Oh, it looks like rustfmt [needs to be run again](https://github.com/tokio-rs/tracing/pull/741/checks?check_run_id=833415009) after the rebase --- mind doing that real quick? Thanks!

LucioFranco commented:
:shipit:

davidbarsky commented:
happy with this.

hawkw commented:
@mockersf sorry to bother you, but I think running rustfmt again is the only remaining blocker to merging this!

mockersf commented:
sorry about the format, I have another project where I have to disable format on save and I keep forgetting to reenable it back after...

fixed now, thanks !

hawkw commented:
Great, thanks! I'm going to merge this branch now, thanks for all your hard work on this!
